graphdatascience/graph/graph_cypher_runner.py

```diff
@@ -0,0 +1,135 @@
+from typing import Any, Optional
+
+from ..query_runner.arrow_query_runner import ArrowQueryRunner
+from ..query_runner.query_runner import QueryRunner
+from ..server_version.server_version import ServerVersion
+from .graph_object import Graph
+from graphdatascience.caller_base import CallerBase
+from graphdatascience.graph.graph_create_result import GraphCreateResult
+
+
+class GraphCypherRunner(CallerBase):
+    def __init__(self, query_runner: QueryRunner, namespace: str, server_version: ServerVersion) -> None:
+        if server_version < ServerVersion(2, 4, 0):
+            raise ValueError("The new Cypher projection is only supported since GDS 2.4.0.")
+        super().__init__(query_runner, namespace, server_version)
+
+    def project(
+        self,
+        query: str,
+        database: Optional[str] = None,
+        **params: Any,
+    ) -> GraphCreateResult:
+        """
+        Run a Cypher projection.
+        The provided query must end with a `RETURN gds.graph.project(...)` call.
+
+        Parameters
+        ----------
+        query: str
+            the Cypher projection query
+        database: str
+            the database on which to run the query
+        params: Dict[str, Any]
+            parameters to the query
+
+        Returns
+        -------
+        A tuple of the projected graph and statistics about the projection
+        """
+
+        GraphCypherRunner._verify_query_ends_with_return_clause(self._namespace, query)
+
+        # See run_cypher
+        qr = self._query_runner
+
+        # The Arrow query runner should not be used to execute arbitrary Cypher
+        if isinstance(qr, ArrowQueryRunner):
+            qr = qr.fallback_query_runner()
+
+        result = qr.run_query(query, params, database, False)
+        result = result.squeeze()
+        try:
+            graph_name = str(result["graphName"])
+        except (KeyError, TypeError):
+            if isinstance(result, str):
+                if result == str(self._server_version):
+                    graph_name = (
+                        "Could not get the graph name from the result. "
+                        "This is probably because this is a unit test and no mock result was given to the query runner"
+                    )
+                else:
+                    # This is likely a test where a mock result was provided
+                    # But it was a dict with a single entry.
+                    # squeeze() will have removed the dict and only left that single value
+                    # so we assume that the key was graphName and use the value.
+                    graph_name = result
+            else:
+                raise ValueError(
+                    f"Invalid query, the query must end with the `RETURN {self._namespace}(...)` call: {query}"
+                )
+
+        return GraphCreateResult(Graph(graph_name, self._query_runner, self._server_version), result)  # type: ignore
+
+    __separators = re.compile(r"[,(.]")
```

Comment:
I'm gonna sneak in some not-top-of-the-class constants in Java as well, I think it makes sense to define those as close to the usage as possible. If it's only uses within one method, I'd define it within this method (if the language permits that) or next to that method, i.e. above. This way, related code is together in the source and I don't have to jump around in the file. If a constant is used across methods, its scope is the whole class and so it can live at the top.
That being said, this is **not** a constant though, this is a class variable, which is supposed to follow the same naming conventions as function and method names.
A constant would be on the module level at the top of the file (and indeed in SCREAMING_UPPER_CASE), but again, I like to reduce the scope of it, so it's not a module level constant because we only need it in the one method.